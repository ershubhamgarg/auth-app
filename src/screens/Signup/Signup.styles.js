import { StyleSheet } from "react-native";
import { LIGHT_BLUE, SECONDARY_BLUE } from "../../constants/styleConstants";

export const styles = StyleSheet.create({
  container: {
    height: "100%",
    padding: 20,
    justifyContent: "center",
  },
  title: { color: LIGHT_BLUE, fontSize: 13, fontWeight: "500" },
  logo: { height: 100, width: 100 },
  button: { marginTop: 50, width: "100%" },
  input: { marginVertical: 15 },
  footer: { marginTop: 10 },
  signup: { color: SECONDARY_BLUE, textDecorationLine: "underline" },
  alignCenter: { alignItems: "center" },
});
