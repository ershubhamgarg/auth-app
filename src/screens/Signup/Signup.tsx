import { useNavigation } from "@react-navigation/native";
import { NativeStackNavigationProp } from "@react-navigation/native-stack";
import { useEffect, useState } from "react";
import {
  Alert,
  Image,
  ImageBackground,
  KeyboardAvoidingView,
  Platform,
  Pressable,
  Text,
  View,
} from "react-native";
import Button from "../../components/Button/Button";
import Input from "../../components/Input/Input";
import { useAuth } from "../../context/AuthContext";
import { HomeStackParamList } from "../../navigation/HomeStack";
import { styles } from "./Signup.styles";
import { ERROR_MSG } from "../../constants/errorConstants";
import { ERROR } from "../../constants/styleConstants";
import { validateEmail } from "../../utils/utils";
export const Signup = () => {
  const [name, setName] = useState("");
  const [nameError, setNameError] = useState("");
  const [email, setEmail] = useState("");
  const [emailError, setEmailError] = useState("");
  const [password, setPassword] = useState("");
  const [passwordError, setPasswordError] = useState("");
  const [confirmPassword, setConfirmPassword] = useState("");
  const [confirmPasswordError, setConfirmPasswordError] = useState("");

  const { signup, user, error, setErrors } = useAuth();
  const navigation =
    useNavigation<NativeStackNavigationProp<HomeStackParamList>>();
  let allGood = true;
  const handleSignup = () => {
    if (!name.length) {
      setNameError(ERROR_MSG.NAME);
      allGood = false;
    }
    if (!email.length) {
      setEmailError(ERROR_MSG.EMAIL);
      allGood = false;
    }
    if (!validateEmail(email)) {
      setEmailError("Please enter a valid email id");
      allGood = false;
    }
    if (!password.length) {
      setPasswordError(ERROR_MSG.PASSWORD);
      allGood = false;
    }
    if (!confirmPassword.length) {
      setConfirmPasswordError(ERROR_MSG.CONFIRM_PASSWORD);
      allGood = false;
    }
    if (
      password.length &&
      confirmPassword.length &&
      password !== confirmPassword
    ) {
      setConfirmPasswordError(ERROR_MSG.PASSWORD_MISMATCH);
      allGood = false;
    }

    if (allGood) {
      const payload = {
        name,
        email,
        password,
      };
      signup(payload);
    }
  };

  useEffect(() => {
    if (user) {
      navigation.navigate("Home");
      return;
    }
    if (error) {
      Alert.alert(error);
      setErrors?.();
    }
  }, [user, error]);

  const onEmailChange = (e: string) => {
    setEmail(e);
    if (e.length) {
      setEmailError("");
    } else {
      setEmailError(ERROR_MSG.EMAIL);
      allGood = false;
    }
  };

  const onPasswordChange = (e: string) => {
    setPassword(e);
    if (e.length) {
      setPasswordError("");
      setConfirmPasswordError("");
    } else {
      setPasswordError(ERROR_MSG.PASSWORD);
      allGood = false;
    }
  };

  const onConfirmPasswordChange = (e: string) => {
    setConfirmPassword(e);
    if (e.length) {
      setConfirmPasswordError("");
    } else {
      setConfirmPasswordError(ERROR_MSG.CONFIRM_PASSWORD);
      allGood = false;
    }
  };

  const onNameChange = (e: string) => {
    setName(e);
    if (e.length) {
      setNameError("");
    } else {
      setNameError(ERROR_MSG.NAME);
    }
  };

  const gotoLogin = () => {
    navigation.goBack();
  };
  return (
    <ImageBackground
      source={require("../../assets/images/back.png")}
      style={styles.container}
    >
      <KeyboardAvoidingView
        behavior={Platform.OS === "ios" ? "padding" : "height"}
      >
        <View style={styles.alignCenter}>
          <Image
            source={require("../../assets/logo/logo.png")}
            style={styles.logo}
          />
          <Text style={styles.title}>Please enter your details to sign up</Text>
        </View>
        <Input
          placeholder="Name"
          label="Name"
          keyboardType="default"
          value={name}
          containerStyle={styles.input}
          onChangeText={onNameChange}
          error={nameError}
        />
        <Input
          placeholder="Email"
          label="Email"
          value={email}
          keyboardType="email-address"
          containerStyle={styles.input}
          onChangeText={onEmailChange}
          error={emailError}
        />
        <Input
          placeholder="Password"
          label="Password"
          isPassword
          value={password}
          containerStyle={styles.input}
          onChangeText={onPasswordChange}
          error={passwordError}
        />
        <Input
          placeholder="Confirm Password"
          label="Confirm Password"
          isPassword
          value={confirmPassword}
          containerStyle={styles.input}
          onChangeText={onConfirmPasswordChange}
          error={confirmPasswordError}
        />

        <Button
          containerStyle={styles.button}
          label="Sign up"
          onPress={handleSignup}
        />
        <View style={styles.alignCenter}>
          <Text style={styles.footer}>
            Already have an account ?{" "}
            <Pressable onPress={gotoLogin}>
              <Text style={styles.signup}>Login</Text>
            </Pressable>
          </Text>
        </View>
      </KeyboardAvoidingView>
    </ImageBackground>
  );
};
