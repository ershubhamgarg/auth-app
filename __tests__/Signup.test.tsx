import React from "react";
import { render, fireEvent, waitFor } from "@testing-library/react-native";
import { Alert } from "react-native";
import { Signup } from "../src/screens/Signup";

// --- Mock navigation ---
const mockNavigate = jest.fn();
const mockGoBack = jest.fn();

jest.mock("@react-navigation/native", () => {
  const actual = jest.requireActual("@react-navigation/native");
  return {
    ...actual,
    useNavigation: () => ({
      navigate: mockNavigate,
      goBack: mockGoBack,
    }),
  };
});

// --- Mock AuthContext ---
const mockSignup = jest.fn();
const mockSetErrors = jest.fn();
let mockUser: any = null;
let mockError: string | null = null;

jest.mock("../src/context/AuthContext", () => ({
  useAuth: () => ({
    signup: mockSignup,
    user: mockUser,
    error: mockError,
    setErrors: mockSetErrors,
  }),
}));

// --- Mock Alert ---
jest.spyOn(Alert, "alert").mockImplementation(jest.fn());

describe("Signup Screen", () => {
  beforeEach(() => {
    jest.clearAllMocks();
    mockUser = null;
    mockError = null;
  });

  test("matches snapshot", () => {
    const { toJSON } = render(<Signup />);
    expect(toJSON()).toMatchSnapshot();
  });

  test("renders all input fields and sign up button", () => {
    const { getByPlaceholderText, getByText } = render(<Signup />);
    expect(getByPlaceholderText("Name")).toBeTruthy();
    expect(getByPlaceholderText("Email")).toBeTruthy();
    expect(getByPlaceholderText("Password")).toBeTruthy();
    expect(getByPlaceholderText("Confirm Password")).toBeTruthy();
    expect(getByText("Sign up")).toBeTruthy();
  });

  test("shows errors when fields are empty on submit", () => {
    const { getByTestId, getByPlaceholderText } = render(<Signup />);
    fireEvent.press(getByTestId("signup-button"));
    const signupContainer = getByTestId("signup-container");
    // Check validation
    expect(signupContainer).toHaveTextContent(/Name is required/);
    expect(signupContainer).toHaveTextContent(/Email is required/);
    expect(signupContainer).toHaveTextContent(/Password is required/);
    expect(signupContainer).toHaveTextContent(/Confirm password is required/);
    expect(mockSignup).not.toHaveBeenCalled();
  });

  test("shows password mismatch error", () => {
    const { getByPlaceholderText, getByText, getByTestId } = render(<Signup />);
    fireEvent.changeText(getByPlaceholderText("Name"), "John");
    fireEvent.changeText(getByPlaceholderText("Email"), "john@example.com");
    fireEvent.changeText(getByPlaceholderText("Password"), "123456");
    fireEvent.changeText(getByPlaceholderText("Confirm Password"), "654321");

    fireEvent.press(getByText("Sign up"));
    const signupContainer = getByTestId("signup-container");
    expect(signupContainer).toHaveTextContent(
      /Password and confirm password do not match/
    );
    expect(mockSignup).not.toHaveBeenCalled();
  });

  test("calls signup when all inputs are valid", () => {
    const { getByPlaceholderText, getByText } = render(<Signup />);
    fireEvent.changeText(getByPlaceholderText("Name"), "John");
    fireEvent.changeText(getByPlaceholderText("Email"), "john@example.com");
    fireEvent.changeText(getByPlaceholderText("Password"), "123456");
    fireEvent.changeText(getByPlaceholderText("Confirm Password"), "123456");

    fireEvent.press(getByText("Sign up"));

    expect(mockSignup).toHaveBeenCalledWith({
      name: "John",
      email: "john@example.com",
      password: "123456",
    });
  });

  test("navigates to Home when user is set", async () => {
    mockUser = { id: 1, name: "John" };
    render(<Signup />);
    await waitFor(() => {
      expect(mockNavigate).toHaveBeenCalledWith("Home");
    });
  });

  test("shows alert when error exists", async () => {
    mockError = "Something went wrong";
    render(<Signup />);
    await waitFor(() => {
      expect(Alert.alert).toHaveBeenCalledWith("Something went wrong");
      expect(mockSetErrors).toHaveBeenCalled();
    });
  });

  test("clears field errors on valid input", () => {
    const { getByPlaceholderText, getByTestId } = render(<Signup />);
    const nameInput = getByPlaceholderText("Name");
    const emailInput = getByPlaceholderText("Email");
    const passwordInput = getByPlaceholderText("Password");
    const confirmPasswordInput = getByPlaceholderText("Confirm Password");

    fireEvent.changeText(nameInput, "Jane");
    fireEvent.changeText(emailInput, "jane@example.com");
    fireEvent.changeText(passwordInput, "abc123");
    fireEvent.changeText(confirmPasswordInput, "abc123");

    const signupContainer = getByTestId("signup-container");
    expect(signupContainer).not.toHaveTextContent(/Name is required/);
    expect(signupContainer).not.toHaveTextContent(/Email is required/);
    expect(signupContainer).not.toHaveTextContent(
      /Please enter a valid email id/
    );
    expect(signupContainer).not.toHaveTextContent(/Email is required/);
    expect(signupContainer).not.toHaveTextContent(
      /Confirm password is required/
    );
    expect(signupContainer).not.toHaveTextContent(
      /Password and confirm password do not match/
    );
  });

  test("navigates back to login when pressing Login text", () => {
    const { getByText } = render(<Signup />);
    fireEvent.press(getByText("Login"));
    expect(mockGoBack).toHaveBeenCalled();
  });
});
