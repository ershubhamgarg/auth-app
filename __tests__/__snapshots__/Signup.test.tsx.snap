// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`Signup Screen matches snapshot 1`] = `
<View
  accessibilityIgnoresInvertColors={true}
  style={
    {
      "height": "100%",
      "justifyContent": "center",
      "padding": 20,
    }
  }
>
  <Image
    source={
      {
        "testUri": "../../../src/assets/images/back.png",
      }
    }
    style={
      [
        {
          "bottom": 0,
          "left": 0,
          "position": "absolute",
          "right": 0,
          "top": 0,
        },
        {
          "height": "100%",
          "width": undefined,
        },
        undefined,
      ]
    }
  />
  <View
    onLayout={[Function]}
    style={
      {
        "paddingBottom": 0,
      }
    }
    testID="signup-container"
  >
    <View
      style={
        {
          "alignItems": "center",
        }
      }
    >
      <Image
        source={
          {
            "testUri": "../../../src/assets/logo/logo.png",
          }
        }
        style={
          {
            "height": 100,
            "width": 100,
          }
        }
      />
      <Text
        style={
          {
            "color": "#16476A",
            "fontSize": 13,
            "fontWeight": "500",
          }
        }
      >
        Please enter your details to sign up
      </Text>
    </View>
    <View
      style={
        [
          {
            "width": "100%",
          },
          {
            "marginVertical": 15,
          },
        ]
      }
    >
      <View
        style={
          {
            "borderBottomWidth": 1,
            "borderColor": "#000000",
            "borderRadius": 4,
            "paddingTop": 10,
            "width": "100%",
          }
        }
      >
        <Text
          style={
            [
              {
                "color": "#878787",
                "fontSize": 12,
                "marginBottom": 5,
              },
              undefined,
            ]
          }
        >
          Name
        </Text>
        <View
          style={
            {
              "flexDirection": "row",
            }
          }
        >
          <TextInput
            autoCapitalize="none"
            autoComplete="off"
            autoCorrect={false}
            keyboardType="default"
            onChangeText={[Function]}
            placeholder="Name"
            secureTextEntry={false}
            style={
              [
                {
                  "fontSize": 16,
                  "paddingVertical": 5,
                },
                {
                  "flex": 1,
                },
              ]
            }
            testID="name-input"
            value=""
          />
        </View>
      </View>
    </View>
    <View
      style={
        [
          {
            "width": "100%",
          },
          {
            "marginVertical": 15,
          },
        ]
      }
    >
      <View
        style={
          {
            "borderBottomWidth": 1,
            "borderColor": "#000000",
            "borderRadius": 4,
            "paddingTop": 10,
            "width": "100%",
          }
        }
      >
        <Text
          style={
            [
              {
                "color": "#878787",
                "fontSize": 12,
                "marginBottom": 5,
              },
              undefined,
            ]
          }
        >
          Email
        </Text>
        <View
          style={
            {
              "flexDirection": "row",
            }
          }
        >
          <TextInput
            autoCapitalize="none"
            autoComplete="off"
            autoCorrect={false}
            keyboardType="email-address"
            onChangeText={[Function]}
            placeholder="Email"
            secureTextEntry={false}
            style={
              [
                {
                  "fontSize": 16,
                  "paddingVertical": 5,
                },
                {
                  "flex": 1,
                },
              ]
            }
            testID="email-input"
            value=""
          />
        </View>
      </View>
    </View>
    <View
      style={
        [
          {
            "width": "100%",
          },
          {
            "marginVertical": 15,
          },
        ]
      }
    >
      <View
        style={
          {
            "borderBottomWidth": 1,
            "borderColor": "#000000",
            "borderRadius": 4,
            "paddingTop": 10,
            "width": "100%",
          }
        }
      >
        <Text
          style={
            [
              {
                "color": "#878787",
                "fontSize": 12,
                "marginBottom": 5,
              },
              undefined,
            ]
          }
        >
          Password
        </Text>
        <View
          style={
            {
              "flexDirection": "row",
            }
          }
        >
          <TextInput
            autoCapitalize="none"
            autoComplete="off"
            autoCorrect={false}
            onChangeText={[Function]}
            placeholder="Password"
            secureTextEntry={true}
            style={
              [
                {
                  "fontSize": 16,
                  "paddingVertical": 5,
                },
                {
                  "flex": 1,
                },
              ]
            }
            testID="password-input"
            value=""
          />
          <View
            accessibilityState={
              {
                "busy": undefined,
                "checked": undefined,
                "disabled": undefined,
                "expanded": undefined,
                "selected": undefined,
              }
            }
            accessibilityValue={
              {
                "max": undefined,
                "min": undefined,
                "now": undefined,
                "text": undefined,
              }
            }
            accessible={true}
            collapsable={false}
            focusable={true}
            onBlur={[Function]}
            onClick={[Function]}
            onFocus={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
          >
            <Image
              resizeMode="contain"
              source={
                {
                  "testUri": "../../../src/assets/icons/hide.png",
                }
              }
              style={
                [
                  {
                    "height": 20,
                    "width": 20,
                  },
                ]
              }
            />
          </View>
        </View>
      </View>
    </View>
    <View
      style={
        [
          {
            "width": "100%",
          },
          {
            "marginVertical": 15,
          },
        ]
      }
    >
      <View
        style={
          {
            "borderBottomWidth": 1,
            "borderColor": "#000000",
            "borderRadius": 4,
            "paddingTop": 10,
            "width": "100%",
          }
        }
      >
        <Text
          style={
            [
              {
                "color": "#878787",
                "fontSize": 12,
                "marginBottom": 5,
              },
              undefined,
            ]
          }
        >
          Confirm Password
        </Text>
        <View
          style={
            {
              "flexDirection": "row",
            }
          }
        >
          <TextInput
            autoCapitalize="none"
            autoComplete="off"
            autoCorrect={false}
            onChangeText={[Function]}
            placeholder="Confirm Password"
            secureTextEntry={true}
            style={
              [
                {
                  "fontSize": 16,
                  "paddingVertical": 5,
                },
                {
                  "flex": 1,
                },
              ]
            }
            testID="confirm-password-input"
            value=""
          />
          <View
            accessibilityState={
              {
                "busy": undefined,
                "checked": undefined,
                "disabled": undefined,
                "expanded": undefined,
                "selected": undefined,
              }
            }
            accessibilityValue={
              {
                "max": undefined,
                "min": undefined,
                "now": undefined,
                "text": undefined,
              }
            }
            accessible={true}
            collapsable={false}
            focusable={true}
            onBlur={[Function]}
            onClick={[Function]}
            onFocus={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
          >
            <Image
              resizeMode="contain"
              source={
                {
                  "testUri": "../../../src/assets/icons/hide.png",
                }
              }
              style={
                [
                  {
                    "height": 20,
                    "width": 20,
                  },
                ]
              }
            />
          </View>
        </View>
      </View>
    </View>
    <View
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": false,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        [
          {
            "alignItems": "center",
            "backgroundColor": "#132440",
            "borderRadius": 5,
            "flexDirection": "row",
            "justifyContent": "center",
            "minHeight": 50,
            "padding": 10,
            "shadowOffset": {
              "height": 1,
              "width": 1,
            },
            "shadowOpacity": 0.4,
          },
          {
            "marginTop": 50,
            "width": "100%",
          },
        ]
      }
      testID="signup-button"
    >
      <Text
        style={
          {
            "color": "#ffffff",
            "fontSize": 15,
            "fontWeight": "700",
            "marginLeft": 10,
          }
        }
      >
        Sign up
      </Text>
    </View>
    <View
      style={
        {
          "alignItems": "center",
        }
      }
    >
      <Text
        style={
          {
            "marginTop": 10,
          }
        }
      >
        Already have an account ?
         
        <View
          accessibilityState={
            {
              "busy": undefined,
              "checked": undefined,
              "disabled": undefined,
              "expanded": undefined,
              "selected": undefined,
            }
          }
          accessibilityValue={
            {
              "max": undefined,
              "min": undefined,
              "now": undefined,
              "text": undefined,
            }
          }
          accessible={true}
          collapsable={false}
          focusable={true}
          onBlur={[Function]}
          onClick={[Function]}
          onFocus={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
        >
          <Text
            style={
              {
                "color": "#3B9797",
                "textDecorationLine": "underline",
              }
            }
          >
            Login
          </Text>
        </View>
      </Text>
    </View>
  </View>
</View>
`;
